### DISCUSSION TITLE
What causes DidChangeConfigurationNotification to fire?

### DISCUSSION THREAD

**Discussion by jdesrosiers** [microsoft/vscode-discussions#1243] [Extension Development QnA] [answered]
I'm writing a language server extension and I'm seeing some unexpected behavior when running it in vscode. When I start the extension, the DidChangeConfigurationNotification gets sent to the server anywhere between zero and three times. This is just starting up the extension, I'm not making changes the settings or anything else.

When should I expect vscode to send a configuration change notification other than when settings change? Should I expect this notification when the extension initializes? Are there any actions the server might be taking that could trigger the change notification to be sent? Could this be a bug in vscode?

This behavior is an issue because settings changes could require the entire workspace to need to be revalidated. That means that handling these notifications is expensive and getting them when nothing is actually changing is problematic.

**Comment by jdesrosiers** [comment]
I just updated vscode and I'm no longer seeing an inconsistency. I'm now consistently seeing the notification one time only when the server starts. It appears this was caused by a bug in vscode and it was fixed in the latest release.

**Reply by jdesrosiers** [reply]
Nevermind, the inconsistency is back. I'm back to getting between 0 and 3 config change notifications when I start the extension.

**Comment by tamuratak** [comment, marked as answer]
If a workspace folder includes `.vscode/settings.json`, the events will be fired. For example, if the workspace-specific setting is loaded after the extension is initialized due to the internal implementation of VS Code, they would be fired. 

Related
- microsoft/vscode/issues/66246

**Reply by jdesrosiers** [reply]
That's really helpful. Thank you! But, it seems like there must be more going on here. That doesn't explain why I'm getting as many as three of these notifications. Today I tried with and without `.vscode/settings.json` in my test workspace and there doesn't seem to be a difference in behavior. I still get between 0 and 3 configuration change notifications. So, there must be other triggers for this notification as well.

> if the workspace-specific setting is loaded after the extension is initialized due to the internal implementation of VS Code, they would be fired.

That's really problematic. So, when I get settings to do an initial validation of the workspace, vscode might not have loaded all the settings yet. Is there any way to know that settings have been fully loaded? Then I can wait for that state before attempting to validate the workspace.

**Reply by jdesrosiers** [reply]
I've been doing some more testing. I setup a test where I set a config to one value in user setting and another value in workspace settings. If I'm running into a race condition regarding loading setting on the client, I should see the settings change after the config change notification is received. At the moment, I'm consistently getting one notification, but it's clear that nothing has changed before and after the notification. So, at least one of the notifications isn't the result either user settings or workspace settings getting loaded. Theoretically, the other two notifications I see sometimes could be each of those types of settings getting loaded, but I haven't been able to confirm that yet.

**Reply by jdesrosiers** [reply]
I got the notification three times again and none of them seem to be from the settings getting loaded or changing at all. `getConfiguration` is returning the same values before, after, and in between all three notifications.

**Reply by tamuratak** [reply]
On my machine, the configuration change events fired during the extension activation are:

- `settingsSync.keybindingsPerPlatform`
- `settingsSync.ignoredExtensions`
- `settingsSync.ignoredSettings`
- `window.newWindowProfile`

You can check by yourself with the [`ConfigurationChangeEvent.affectsConfiguration`](https://code.visualstudio.com/api/references/vscode-api#ConfigurationChangeEvent) method. Because they are special settings for the settings, it would be reasonable they are being fired.

If you think it is a bug, you had better report it to https://github.com/Microsoft/vscode

**Reply by jdesrosiers** [reply]
Thanks, this helped me figure it out! The number of ConfigurationChangeEvents in the extension correlates perfectly to the number of DidChangeConfigurationNotifications I'm receiving in the language server. I'm not getting exactly the same events as you, but everything other than `window.newWindowProfile` triggers a config change notification to the language server. I'm not sure what all of the config changes are other than that they aren't for my extension. I couldn't tell that from the language server notification because all I get is `{ "settings": null }`. It's intended to just notify that _something_ changed. It seems like a bug for it to notify when something unrelated to the language server changes, but maybe that's on purpose for some reason. It's certainly unhelpful for there to not even be a way to know that what changed isn't related to the language server.

Anyway, the fix was to change this,

```typescript
connection.client.register(DidChangeConfigurationNotification.type, undefined);
```

to this,

```typescript
connection.client.register(DidChangeConfigurationNotification.type, { section: "myLanguageServer" });
```

The latter is supposed to be deprecated, but it solves the problem. I now only get notifications if the configs I care about change.
